### ISSUE TITLE
Separate locale file for each module or multiple locale files

### ISSUE THREAD

**Issue by 3adel-bassiony** [QuiiBz/next-international#267]
**Is your feature request related to a problem? Please describe.**
When using next-international in a big project with support for multiple languages it's hard to add all locales in one file, so I think it will be better if we have an option to divide the locale file into multiple locale files, so each feature or module can have a separate locale file.

**Describe the solution you'd like**
Divide the locale files into multiple ts files so we can use a different locale file for each module

**Comment by QuiiBz** [comment]
Indeed, that's currently one of the limitation of next-international.

> so we can use a different locale file for each module

Could you describe what "module" means in this context? Are you referring to a JavaScript module?

**Comment by irfancoder** [comment]
> Could you describe what "module" means in this context? Are you referring to a JavaScript module?

I think he is referring to namespaces, like in [i18next.](https://www.i18next.com/principles/namespaces) In my previous work, we group the translations by route, and load the namespaces as necessary. I agree with OP this is a very much a needed feature +1

**Comment by Marsony** [comment]
Here is a minimal repro (with niche use case):

create-i18n.ts:
```
type Dictionary = {
  [key: string]: () => Promise<any>;
};

type LocaleList = {
  [key: string]: Dictionary;
};

type ComposeDictionaries<D extends LocaleList, K extends keyof D = keyof D> = {
  [P in keyof D[K]]: Awaited<ReturnType<D[K][P]>>['default'];
};

type IntermediateType<Locales extends LocaleList> = ComposeDictionaries<Locales>;

export const createI18n = <Locales extends LocaleList>({ locales }: { locales: Locales }) => {
  type MergedDict = IntermediateType<Locales>;
  type Locale = FlattenLocale<MergedDict>;

  const I18nContext = createContext<I18nProviderContext<Locale> | null>(null);

  const useI18n = createUseI18n(I18nContext);
  ...
```

get-translations.ts (use only in the next server-side methods):
```
import type { BaseLocale } from 'next-international';

import { i18n } from '../i18n';

const flatLocale = <Locale extends BaseLocale>(
  locale: Record<string, unknown>,
  prefix = '',
): Locale =>
  Object.entries(locale).reduce(
    (prev, [name, value]) => ({
      ...prev,
      ...(typeof value === 'string'
        ? { [prefix + name]: value }
        : flatLocale(value as unknown as Locale, `${prefix}${name}.`)),
    }),
    {} as Locale,
  );

export const getTranslations = ({ language, url }: { language: string; url: string }) => {
  const { pages } = i18n;

  const toLoad = [
    ...pages['*'],
    ...pages[url],
  ].map((dictionary) =>
    import(`../dictionaries/${language}/${dictionary}`).then((c) => ({
      [dictionary]: c.default,
    })),
  );

  return Promise.all(toLoad).then((r) => {
    let s: BaseLocale = {};

    for (const d of r) {
      const key = Object.keys(d)[0];
      s = { ...s, ...flatLocale(d[key], key + '.') };
    }

    return s;
  });
};
```

i18n.ts:
```
export const i18n = {
  pages: {
    '*': ['auth', 'navbar'],
  },
};
```

index.ts:
```
import { createI18n } from './create-i18n';

export const { I18nProvider, useI18n, useScopedI18n } = createI18n({
  locales: {
    en: {
      auth: () => import('./dictionaries/en/auth'),
      navbar: () => import('./dictionaries/en/navbar'),
    },
  },
});
```

Please note that the example given is a custom wrapper over the createT method of this library. These are just ideas that you can apply if you want to achieve a similar effect.

**Comment by 3adel-bassiony** [comment]
@irfancoder Yes that's correct, I mean something like Namespaces in i18next
@QuiiBz In this case I can divide the locale file into multiple files and work on them easily

**Comment by QuiiBz** [comment]
Thanks for the details! This feature certainly translates to APIs breaking change/new APIs to support this behaviour. I'll take a look at how other libraries have designed this feature, so feel free to also share your thoughts!

**Comment by olafurns7** [comment]
@QuiiBz just to give you a term of scope, in our use case we have ~7000 translation strings, in 14 locales :)

**Comment by chungweileong94** [comment]
This is kinda related to this issue/question #180. 

Wrote a simple repo myself to test out this library, and this is what I found:
- Bundle size is not an issue on server component, but it might be an issue while loading huge locale files? 🤷
- Bundle size will be an issue on client component, all the locale files for the selected language will be loaded.

**Comment by aabmets** [comment]
@3adel-bassiony @QuiiBz 
Do you mean something like this?
![image](https://github.com/QuiiBz/next-international/assets/6948036/06dd51b6-f1b2-4b77-84df-7f8d261109b9)
![image](https://github.com/QuiiBz/next-international/assets/6948036/43b30c8d-f8eb-4c3b-8a1c-c404e8958f48)
![image](https://github.com/QuiiBz/next-international/assets/6948036/0a22c467-523b-48fa-818f-f0fc425ae7f5)

**Comment by QuiiBz** [comment]
@aabmets not really, because we want to only load the necessary locale file when needed. If you split your locales per page, you only want to load the locales for this page.

**Comment by aabmets** [comment]
@QuiiBz is it possible to define all localizations in one place (like the translations folder i have) and then implement this feature, that next-i11l would be able to only import a subset of these localization files with `const t = await getI18n(['ns1', 'ns2'])`?

**Comment by QuiiBz** [comment]
Not right now, but that's the goal of this issue, yes - but the API would probably be different, and in order to keep loading only the necessary file, you wouldn't have any `index.ts` which returns every locale file.

**Comment by Meierschlumpf** [comment]
Not sure about the performance, but maybe something like this would already be enough:
![image](https://github.com/QuiiBz/next-international/assets/63781622/d7cfea5f-15a6-46fb-b21d-3a10b702ed50)

I've got full typesafety for the defined namespace. But of course I'm also creating a new instance of I18nClient whenever I call the useI18n method

**Comment by Meierschlumpf** [comment]
Nevermind It's more complicated :joy:

**Comment by gustaveWPM** [comment]
The more I think about it, the more I think we should be able to write the object of a locale with both inlined translations and lazy-loaded fragments.
